Yosemite/Yosemite/Stores/ReceiptStore.swift

```diff
@@ -69,31 +87,52 @@ private extension ReceiptStore {
     }
 
     func generateCartTotals(order: Order, parameters: CardPresentReceiptParameters) -> [ReceiptTotalLine] {
-        var totalLines = [ReceiptTotalLine]()
-        if NSDecimalNumber(apiAmount: order.totalTax).decimalValue > 0.00 {
-            totalLines.append(ReceiptTotalLine(description: ReceiptContent.Localization.totalTaxLineDescription,
-                                          amount: order.totalTax))
-        }
-        totalLines.append(ReceiptTotalLine(description: ReceiptContent.Localization.amountPaidLineDescription, amount: parameters.formattedAmount))
-        return totalLines
+        let subtotalLines = [discountLine(order: order),
+                             lineIfNonZero(description: ReceiptContent.Localization.shippingLineDescription,
+                                           amount: order.shippingTotal),
+                             lineIfNonZero(description: ReceiptContent.Localization.totalTaxLineDescription,
+                                           amount: order.totalTax)]
+            .compactMap { $0 }
+        let totalLine = [ReceiptTotalLine(description: ReceiptContent.Localization.amountPaidLineDescription,
+                                         amount: parameters.formattedAmount)]
+
+        return subtotalLines + totalLine
     }
 
-    func generateReceiptContent(order: Order, parameters: CardPresentReceiptParameters) -> ReceiptContent {
-        let lineItems = generateLineItems(order: order)
-        let cartTotals = generateCartTotals(order: order, parameters: parameters)
+    func discountLine(order: Order) -> ReceiptTotalLine? {
+        let discountValue = NSDecimalNumber(apiAmount: order.discountTotal).decimalValue
+        if discountValue == 0 && order.coupons.isEmpty {
+            return nil
+        }
+        return ReceiptTotalLine(description: discountLineDescription(order: order),
+                                amount: discountLineAmount(order: order, value: discountValue))
+    }
 
-        return ReceiptContent(parameters: parameters, lineItems: lineItems, cartTotals: cartTotals)
+    func discountLineDescription(order: Order) -> String {
+        var couponCodes = ""
+        if order.coupons.count > 0 {
+            couponCodes = order.coupons.map {
+                $0.code
+            }
+            .joined(separator: ", ")
+            couponCodes = "(\(couponCodes))"
+        }
+        return String.localizedStringWithFormat(ReceiptContent.Localization.discountLineDescription, couponCodes)
     }
 
-    func print(order: Order, parameters: CardPresentReceiptParameters, completion: @escaping (PrintingResult) -> Void) {
-        let content = generateReceiptContent(order: order, parameters: parameters)
-        receiptPrinterService.printReceipt(content: content, completion: completion)
+    func discountLineAmount(order: Order, value: Decimal) -> String {
+        if value > 0 {
```

Comment:
This seems a bit problematic. In a previous check, we don't show this if it's zero, and I don't think WooCommerce will let you create a coupon with a negative discount. That said, if that ever happens, both positive and negative `discountTotal` values would show a `-` sign. I think we should negate the numeric value before converting to string instead.

Reply:
Yeah, I'm not super happy with the treatment of numbers here either and I've got a task to look in to how we can do it better. If we're going to convert the numbers from what the API gives us for display, I want to be really careful about how we do it. Do you know of any other examples in the app where we already do that?

It also probably ties in to getting the `CurrencyFormatter` available here, which is another thing on my list to look in to.

My reasoning for this code is:
1. The common case we want to cover is making a positive API value negative,
2. We don't want a zero value to be negative, because that's weird.
3. It seems like negative API values for discount aren't really a thing.

But that isn't super robust so I agree we should do something about point 3, so we don't confuse things. Most likely example I can think of for a positive discount is perhaps on a refund receipt. I'm going to add an issue to look at this after I have investigated numbers as mentioned above, but given we can't find a current use case where positive discounts exist, I think it's best to merge this for now.